pip/pip-385.md

```diff
@@ -0,0 +1,398 @@
+            Client-xServer: Send(3, message3)
+            Note over Client, Server: After some time
+            opt topic/connection unthrottled
+                note right of Server: TCP read resumed
+                deactivate Server
+                Server-->>Server: read message 3
+                Client->>Server: Send(4, message4)
+            end
+        end
+    end
+    Server-->>Client: SendReceipt(2,msgId2)
+    Server-->>Client: SendReceipt(3,msgId3)
+    Server-->>Client: SendReceipt(4,msgId4)
+
+    Client->>Server: CloseProducer(producerId, reqId)
+    Server-->>-Client: Success(reqId)
+    deactivate Client
+```
+
+## Java client changes
+
+* Client will now have logic to understand the `ThrottleProducer` command and take relevant action of blocking further
+  messages for the relevant topic. It will then respond back with `ThrottleProducerReceipt` command.
+    * Client will resume message sending after the specified time in the `ThrottleProducer` command's data.
+    * This interval of no messages will be noted as "being throttled"
+    * Within this duration, another `ThrottleProducer` command from server may come.
+* Producer will record new OTel metric indicating which topic was throttled and the reason.
+* In case a message fails due to timeout and there was a throttled command from server for the owning topic, client will
+  instead throw a rate limit exception instead of timeout exception.
+
+# Detailed Design
+
+## High-level Implementation Details
+
+### Broker Changes
+
+* For calls arising from `PublishRateLimiterImpl` class, add logic in `ServerCnxThrottleTracker.java` to send the
+  command to client and wait for response for the max configured duration before calling `changeAutoRead`. It checks for
+  feature availability first.
+* For calls arising from `ServerCnxThrottleTracker::changeThrottlingFlag`, we send the command async (if feature
+  supported) without worrying about response and then call `changeAutoRead`.
+* capture `AbstractTopic::getTotalPublishRateLimitCounter` per publish rate limit counter and add relevant attribute in
+  the rate limit metric
+
+### Determining the throttling duration for clients
+
+Broadly, there are 2 different categories for rate limiting - (a) quota exceeding at a topic/resource group level OR (b)
+connection level breaches leading to broker's resource crunch.
+
+In case (a), as per the current token bucket based quota tracking introduced in PIP-322, we (lazily) resume channel
+reads on the next second mark. i.e. the throttling is only applied at max for 1 second.
+In case (b), there is no clear picture on when the channel read is resumed as we wait till we have processed half of the
+pending messages, or cleared half of the memory threshold.
+
+Thus, for case (a), we will be asking the client to pause further message sending for the remainder of the second and in
+case (b), we will send `0` as the channel read is anyway paused.
```

Comment:
This doesn't make sense. The previous comment starting "This logic might need some more careful thought." contains more details about the problem to solve. The solution should take all ways how the broker might throttle producing and calculate a single throttling period. 

In the previous comment, I suggested frequently calculating the RTT of the client-broker connection using Ping/Pong commands and sending the throttle producer command early enough so that TCP level throttling (switching auto read to false) wouldn't have to take place. The similar solution should be expanded to cover all ways that producers get rate limited. For the broker level throttling, a similar approach could be used. My assumption is that by using the up-to-date RTT information, the broker could predict when the throttling would be needed and send it out considering the lag that there is until the command takes effect. 

Besides RTT calculation, It could be useful to add a timestamp field to the CommandPing and CommandPong messages besides the request_id field since that information could be used to also estimate the clock differences between the client and the broker. The timestamps might be useful in calculating how long it takes for a message to arrive on the broker since it was created in the client. 

I'm getting more confident that the problem is solvable, but we need more design and experimentation to get it solved. Ignoring this problem would make this PIP not very useful since it wouldn't meet the goal of avoiding the TCP level throttling. That cannot be disabled completely since the broker will quickly run out of memory at high publish rates. I hope that the problem is clear and we are aligned to address it as part of this PIP.

Reply:
> Is this comment specifically about the case b?

yes

Reply:
While I understand that the way broker is handling the situation is different in case b than case a, but from a protocol perspective, there is no difference. In both cases, client gets the `ProducerThrottled` command and sends back receipt. Client is again expected to stop producing further.
 
case b can happen due to either memory consumption per connection and pending message count per connection exceeding the configured value via the configs `maxMessagePublishBufferSizeInMB ` and `maxPendingPublishRequestsPerConnection` respectively. Both of these situations have a direct impact on broker's performance right away and are not _just_ the cause of noisy neighbour to other topics sharing the connection.

Even if we effectively calculate RTT for the connection, we are still left with the challenge to correctly "predict" when the memory would exceed. message count could be easier to do but since this is affecting multiple topics, the message size pattern could be very different across those topics. Thus, in any situation, if we predict when the overall connection would exceed the memory footprint, it would either undershoot or overshoot the limit.

Taking a step back, for case a, we are actually introducing the comment so that we _specifically_ do not _need_ to pause the channel so that the non-breaching topics may continue to produce. In case b, this need doesn't exist.. so I do not see why we would still like to keep the channel open for read? Do you see any specific reason to let the channel read in resumed state and wait for all topics of the connection to send back a throttle receipt? Specially we may be already exceeding physical resources.